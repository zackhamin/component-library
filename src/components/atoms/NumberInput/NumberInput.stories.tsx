import { Story } from '@storybook/react';
import { NumberInput, NumberInputProps } from './NumberInput';

export default {
	title: 'atoms/NumberInput',
	component: NumberInput,
	argTypes: { onChange: { action: 'changed' } },
};

const Template: Story<NumberInputProps> = (args) => <NumberInput {...args} />;

const defaultProps = {
	min: 1,
	max: 10,
	value: 1,
	id: 'numberId',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Disabled = Template.bind({});
Disabled.args = {
	...defaultProps,
	disabled: true,
};
