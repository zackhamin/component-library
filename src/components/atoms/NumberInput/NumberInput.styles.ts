import styled from 'styled-components';
import { colors, disabledStyles } from '../../../styles';

const StyledNumberWrapper = styled.div<{
	disabled?: boolean;
}>`
	button {
		background-color: ${colors.white};
		color: ${colors.emerald};
		line-height: 2;
		padding: 0 14px;
		font-weight: 800;
		font-size: 1.6rem;
		border: 1px solid ${colors.lightGrey};
		border-radius: 4px;

		:disabled {
			${disabledStyles}
		}
	}
`;

const StyledNumberInput = styled.input<{
	disabled?: boolean;
}>`
	padding: 0.7rem 1.5rem;
	margin: 0 5px 0;
	border: none;
	border-radius: 4px;
	text-align: center;
	::-webkit-outer-spin-button,
	::-webkit-inner-spin-button {
		-webkit-appearance: none;
		margin: 0;
	}
	-moz-appearance: textfield;
	${({ disabled }) => disabled && ` ${disabledStyles}`}
`;

export { StyledNumberInput, StyledNumberWrapper };
