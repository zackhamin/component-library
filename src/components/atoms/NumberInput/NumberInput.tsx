import React, { useEffect, useLayoutEffect, useRef, useState } from 'react';
import { StyledNumberInput, StyledNumberWrapper } from './NumberInput.styles';

export interface NumberInputProps {
	min?: number;
	max?: number;
	value?: number;
	id: string;
	testId?: string;
	disabled?: boolean;
	onChange: (qty: number) => void;
}

const NumberInput: React.FC<NumberInputProps> = ({
	min = 0,
	max = 30,
	value,
	onChange,
	id,
	testId,
	disabled,
}: NumberInputProps) => {
	const [numberValue, setNumberValue] = useState(value || min);
	const firstRender = useRef(true);

	useLayoutEffect(() => {
		if (firstRender.current) {
			firstRender.current = false;
			return;
		}

		onChange(numberValue);
	}, [numberValue]);

	useEffect(() => {
		if (value !== undefined) {
			setNumberValue(value);
		}
	}, [value]);

	const handleDecrease = () => setNumberValue((prev) => prev - 1);
	const handleIncrease = () => setNumberValue((prev) => prev + 1);
	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const inputValue = Number(e.target.value);
		if (isNaN(inputValue)) {
			setNumberValue((prev) => prev);
		} else if (inputValue > max) {
			setNumberValue(max);
		} else if (inputValue < min) {
			setNumberValue(min);
		} else {
			setNumberValue(inputValue);
		}
	};

	return (
		<StyledNumberWrapper>
			<button
				aria-describedby={id}
				type="button"
				role="button"
				onClick={handleDecrease}
				disabled={disabled || numberValue === min}
			>
				&#8211;
			</button>
			<StyledNumberInput
				type="number"
				disabled={disabled}
				role="input"
				data-testid={testId}
				min={min}
				max={max}
				id={id}
				value={numberValue}
				onChange={handleChange}
			/>
			<button
				aria-describedby={id}
				type="button"
				role="button"
				onClick={handleIncrease}
				disabled={disabled || numberValue === max}
			>
				+
			</button>
		</StyledNumberWrapper>
	);
};

export { NumberInput };
