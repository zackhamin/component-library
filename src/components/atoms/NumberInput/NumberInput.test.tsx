import { NumberInput } from '../NumberInput';
import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps = {
	id: 'idProp',
	onChange: jest.fn(),
};

describe('NumberInput', () => {
	it('renders correctly', () => {
		const { container } = render(<NumberInput {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});
	it('has increment and decrement buttons', () => {
		render(<NumberInput {...defaultProps} />);
		expect(screen.getByText('+')).toBeInTheDocument();
		expect(screen.getByText('â€“')).toBeInTheDocument();
		expect(screen.getAllByRole('button')).toHaveLength(2);
	});
	it('has input with "0" default value', () => {
		render(<NumberInput {...defaultProps} />);
		expect(screen.getByRole('input')).toBeInTheDocument();
		expect(screen.getByRole('input')).toHaveDisplayValue('0');
	});
	it('has default id prop', () => {
		render(<NumberInput {...defaultProps} />);
		const id = screen.getByRole('input').getAttribute('id');
		expect(id).toBe('idProp');
	});
});

describe('props', () => {
	it('can add additional prop', () => {
		render(
			<NumberInput
				{...defaultProps}
				value={10}
				max={20}
				min={1}
				testId={'testId'}
			/>
		);
		const maxProps = screen.getByRole('input').getAttribute('max');
		const minProps = screen.getByRole('input').getAttribute('min');
		expect(maxProps).toBe('20');
		expect(minProps).toBe('1');
		expect(screen.getByRole('input')).toHaveDisplayValue('10');
	});
});

describe('Number Input to have a disabled state', () => {
	it('renders with disabled state', () => {
		render(<NumberInput {...defaultProps} disabled={true} />);
		expect(screen.getByRole('input')).toHaveAttribute('disabled');
	});
	it('renders max button with disabled state', () => {
		render(<NumberInput {...defaultProps} max={10} value={10} />);
		expect(screen.getAllByRole('button')[1]).toHaveAttribute('disabled');
	});
	it('renders min button with disabled state', () => {
		render(<NumberInput {...defaultProps} min={10} value={10} />);
		expect(screen.getAllByRole('button')[0]).toHaveAttribute('disabled');
	});
});

describe('button click handlers', () => {
	it('fires change events on change of value', () => {
		const handleOnChange = jest.fn();

		render(<NumberInput {...defaultProps} onChange={handleOnChange} />);
		const input = screen.getByRole('input');
		fireEvent.change(input, { target: { value: '23' } });
		expect(handleOnChange).toHaveBeenCalled();
	});
});
