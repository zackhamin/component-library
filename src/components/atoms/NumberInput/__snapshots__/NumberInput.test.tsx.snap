// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders correctly 1`] = `
.c0 button {
  background-color: #ffffff;
  color: #00594C;
  line-height: 2;
  padding: 0 14px;
  font-weight: 800;
  font-size: 1.6rem;
  border: 1px solid #EEEEEF;
  border-radius: 4px;
}

.c0 button:disabled {
  opacity: 0.3;
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: default;
}

.c1 {
  padding: 0.7rem 1.5rem;
  margin: 0 5px 0;
  border: none;
  border-radius: 4px;
  text-align: center;
  -moz-appearance: textfield;
}

.c1::-webkit-outer-spin-button,
.c1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-describedby="idProp"
      disabled=""
      role="button"
      type="button"
    >
      â€“
    </button>
    <input
      class="c1"
      id="idProp"
      max="30"
      min="0"
      role="input"
      type="number"
      value="0"
    />
    <button
      aria-describedby="idProp"
      role="button"
      type="button"
    >
      +
    </button>
  </div>
</div>
`;
